### ISSUE TITLE
It is not currently possible to enable experiments for package dependencies in the analyzer 

### ISSUE THREAD

**Issue by jakemac53** [dart-lang/sdk#55688]
This comes up for macros - it is not possible today to release an experimental macro package and have people try it out.

In my understanding, this is because analysis_options.yaml files are only loaded if they are in the workspace. So, even if the package has an analysis_options.yaml which enables the experiment, it won't be actually enabled unless you add that package to your workspace.

This is contrary to how other tools work, because they globally enable experiments.

**Comment by jakemac53** [comment]
> when the analyzer is analyzing code in or reachable from a file in the package containing that options file

From what I understand, this is not how it works. If your package enables the experiment via analysis_options.yaml, it _doesn't_ enable it for transitively reachable files (specifically, we care about package dependencies in this case).

If it can be changed to work that way, that would resolve the issue. Or if I am mistaken and it does work that way, but then I don't understand what was happening in the original issue.

> One work-around for users is just to require that if they want to depend on a package that requires an experiment (P2), then they need to enable that experiment in the analysis options file (just as they would if compiling from the command-line) of the root package (P1). That doesn't feel unreasonable to me, but YMMV.

This is exactly the UX I want (enable experiment in my app, and all my dependencies also get that experiment enabled). 

To me that sounds the closest to how the other tools work.

**Comment by bwilkerson** [comment]
> > when the analyzer is analyzing code in or reachable from a file in the package containing that options file
>
> From what I understand, this is not how it works. If your package enables the experiment via analysis_options.yaml, it doesn't enable it for transitively reachable files (specifically, we care about package dependencies in this case).

@scheglov Since I might have misunderstood.

If the experiments in a package being analyzed are not enabled when analyzing files from depended on packages (when they're being analyzed in the original package's context): Why not? and Can they be?

**Comment by scheglov** [comment]
IIUC, with changes that @pq implemented, there are no such thing as "experiments in a package".
There are `analysis_options.yaml` file that govern analysis of some Dart files.

I propose that we bring this idea to its logical conclusion, and use `analysis_options.yaml` files in dependencies too.

**Comment by bwilkerson** [comment]
For all of the other analysis options, I think it makes sense to allow them to be specified per file. I'm not as convinced that it's helpful to users to be able to have different experiments enabled for different files within a package. Of course, doing so would answer the question of what it means if an options file below the root of the package enables experiments.

> I propose that we bring this idea to its logical conclusion, and use `analysis_options.yaml` files in dependencies too.

For everything other than experiments I think that would be a no-op. I don't think we're going to enable lints, apply severity changes, or any other options that I can think of when analyzing those files.

And, it would have a (possibly small) performance implication, because now we'd need to search the directory structure within these packages in order to locate the `analysis_options.yaml` files. It's not clear that the cost would be worth the benefit.

**Comment by scheglov** [comment]
A user should rarely enable experiments at all. But if he does, I think it is fine for us to implement it the most straightforward way. Here this means using `analysis_options.yaml` that corresponds to the file. Consistent, and easy to understand.

> And, it would have a (possibly small) performance implication, because now we'd need to search the directory structure within these packages in order to locate the `analysis_options.yaml` files. It's not clear that the cost would be worth the benefit.

We will have to walk to the root only once.
Which looks cheap to me.

**Comment by bwilkerson** [comment]
> We will have to walk to the root only once.

I don't know what that means.

If a package under development depends on another package, and imports N different libraries from that other package, won't we have to find the analysis options file associated with each of those N libraries? Which seems to me like we're either walking up to the root of the package N times or we're traversing the whole package structure one time.

**Comment by scheglov** [comment]
> > We will have to walk to the root only once.
> 
> I don't know what that means.
> 
> If a package under development depends on another package, and imports N different libraries from that other package, won't we have to find the analysis options file associated with each of those N libraries? Which seems to me like we're either walking up to the root of the package N times or we're traversing the whole package structure one time.

Each directory has only one `analysis_options.yaml` that controls it. So, in case of a single package with `N` libraries we will go to the package root only once, the next time we will go up only to `lib` :-) Or a child of `lib`. We will do `N` file system probes only if each of `N` libraries is in its own directory.

In the worst case, if the package does not have its own `analysis_options.yaml`, we will walk to the file system root, once.

This presumes that we use some options file producing class with caching while building analysis contexts.

**Comment by bwilkerson** [comment]
> In the worst case, if the package does not have its own `analysis_options.yaml`, we will walk to the file system root, once.

For any package inside `.pub-cache` I think we could stop at the root of the package. I don't think we'd want to use any options that weren't downloaded with the package.

**Comment by jakemac53** [comment]
Is anybody working on this one? We might want to consider bumping the priority of it. We have a number of experiments with macros out there (https://pub.dev/packages/freezed/versions/3.0.0-0.0.dev) as an example.

I don't believe any of these can work today, unless you are developing within the package itself, or otherwise have it open in a workspace.

**Comment by bwilkerson** [comment]
> Is anybody working on this one?

I believe that the answer is no. This is still on our radar, but is lower priority than some other work at the moment.
